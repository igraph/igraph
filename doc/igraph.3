.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\"
.\" Copyright (C) 2006-2021  The igraph development team
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2, or (at your option) any later
.\" version.
.\"
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License with
.\" your Debian GNU/Linux system, in /usr/share/common-licenses/GPL, or with
.\" the dpkg source package as the file COPYING.  If not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH IGRAPH 3 "May 2021" "igraph library"
.SH NAME
igraph \- a library for creating and manipulating graphs
.SH DESCRIPTION
.B igraph
is a C library for complex network analysis and graph theory, with emphasis on
efficiency, portability and ease of use.
.SH DOCUMENTATION
The full documentation can be downloaded from the homepage of the
library:
.RI < https://igraph.org/c/doc >
.SH BUGS
If you think you have found a bug in igraph, feel free to file a bug report
in the issue tracker at:
.RI < https://github.com/igraph/igraph/issues >

.SH AUTHORS
Gabor Csardi <csardi.gabor@gmail.com>,
.br
Tamas Nepusz <ntamas@gmail.com>,
.br
Vincent Traag <vtraag@gmail.com>,
.br
Szabolcs Horvat <szhorvat@gmail.com>,
.br
Fabio Zanini <fabio.zanini@fastmail.fm>,
.br
Daniel Noom <ggatw@outlook.com>
