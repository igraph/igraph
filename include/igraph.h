/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2003-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_H
#define IGRAPH_H

#ifndef _GNU_SOURCE
    #define _GNU_SOURCE 1
#endif

#include "igraph_version.h"
#include "igraph_memory.h"
#include "igraph_error.h"
#include "igraph_random.h"
#include "igraph_progress.h"
#include "igraph_statusbar.h"

#include "igraph_types.h"
#include "igraph_complex.h"
#include "igraph_vector.h"
#include "igraph_matrix.h"
#include "igraph_array.h"
#include "igraph_dqueue.h"
#include "igraph_stack.h"
#include "igraph_heap.h"
#include "igraph_psumtree.h"
#include "igraph_strvector.h"
#include "igraph_vector_ptr.h"
#include "igraph_spmatrix.h"
#include "igraph_sparsemat.h"
#include "igraph_qsort.h"

#include "igraph_constants.h"
#include "igraph_datatype.h"
#include "igraph_iterators.h"
#include "igraph_interface.h"
#include "igraph_constructors.h"
#include "igraph_games.h"
#include "igraph_microscopic_update.h"
#include "igraph_centrality.h"
#include "igraph_paths.h"
#include "igraph_components.h"
#include "igraph_structural.h"
#include "igraph_transitivity.h"
#include "igraph_neighborhood.h"
#include "igraph_topology.h"
#include "igraph_bipartite.h"
#include "igraph_cliques.h"
#include "igraph_layout.h"
#include "igraph_visitor.h"
#include "igraph_community.h"
#include "igraph_conversion.h"
#include "igraph_foreign.h"
#include "igraph_motifs.h"
#include "igraph_operators.h"
#include "igraph_flow.h"
#include "igraph_nongraph.h"
#include "igraph_cocitation.h"
#include "igraph_adjlist.h"
#include "igraph_attributes.h"
#include "igraph_blas.h"
#include "igraph_lapack.h"
#include "igraph_arpack.h"
#include "igraph_mixing.h"
#include "igraph_separators.h"
#include "igraph_cohesive_blocks.h"
#include "igraph_eigen.h"
#include "igraph_hrg.h"
#include "igraph_threading.h"
#include "igraph_interrupt.h"
#include "igraph_scg.h"
#include "igraph_matching.h"
#include "igraph_embedding.h"
#include "igraph_scan.h"
#include "igraph_graphlets.h"
#include "igraph_epidemics.h"
#include "igraph_lsap.h"
#include "igraph_coloring.h"

#endif
