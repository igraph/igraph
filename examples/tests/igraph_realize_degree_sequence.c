
#include <igraph.h>
#include <stdio.h>

#include "test_utilities.inc"

int main() {
    igraph_t graph;

    {
        igraph_vector_t ds;
        const igraph_real_t rawds[] = { 3, 2, 2, 1 };
        igraph_vector_view(&ds, &rawds[0], sizeof(rawds) / sizeof(igraph_real_t));

        printf("\n");
        print_vector_round(&ds, stdout);
        printf("\n");

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_LARGEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_SMALLEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_INDEX);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);
    }

    {
        igraph_vector_t ds;
        const igraph_real_t rawds[] = {1, 3, 3, 4, 1, 2, 1, 1, 1, 3};
        igraph_vector_view(&ds, &rawds[0], sizeof(rawds) / sizeof(igraph_real_t));

        printf("\n");
        print_vector_round(&ds, stdout);
        printf("\n");

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_LARGEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_SMALLEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_INDEX);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);
    }

    {
        igraph_vector_t ds;
        const igraph_real_t rawds[] = {2, 0, 3, 2, 2, 2, 2, 3};
        igraph_vector_view(&ds, &rawds[0], sizeof(rawds) / sizeof(igraph_real_t));

        printf("\n");
        print_vector_round(&ds, stdout);
        printf("\n");

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_LARGEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_SMALLEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ds, NULL, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_INDEX);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);
    }

    {
        igraph_vector_t ods, ids;
        const igraph_real_t rawods[] = {3, 0, 1, 1, 1, 1, 0, 1};
        const igraph_real_t rawids[] = {2, 1, 0, 2, 2, 1, 0, 0};
        igraph_vector_view(&ods, &rawods[0], sizeof(rawods) / sizeof(igraph_real_t));
        igraph_vector_view(&ids, &rawids[0], sizeof(rawids) / sizeof(igraph_real_t));

        printf("\n");
        print_vector_round(&ods, stdout);
        print_vector_round(&ids, stdout);
        printf("\n");

        igraph_realize_degree_sequence(&graph, &ods, &ids, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_LARGEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ods, &ids, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_SMALLEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ods, &ids, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_INDEX);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);
    }

    {
        igraph_vector_t ods, ids;
        const igraph_real_t rawods[] = {3, 1, 2, 3, 1, 2, 2};
        const igraph_real_t rawids[] = {2, 2, 1, 2, 3, 2, 2};
        igraph_vector_view(&ods, &rawods[0], sizeof(rawods) / sizeof(igraph_real_t));
        igraph_vector_view(&ids, &rawids[0], sizeof(rawids) / sizeof(igraph_real_t));

        printf("\n");
        print_vector_round(&ods, stdout);
        print_vector_round(&ids, stdout);
        printf("\n");

        igraph_realize_degree_sequence(&graph, &ods, &ids, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_LARGEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ods, &ids, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_SMALLEST);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);

        igraph_realize_degree_sequence(&graph, &ods, &ids, IGRAPH_SIMPLE_SW, IGRAPH_REALIZE_DEGSEQ_INDEX);
        print_graph(&graph, stdout);
        igraph_destroy(&graph);
    }

    return 0;
}
