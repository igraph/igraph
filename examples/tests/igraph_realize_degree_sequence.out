( )

SIMPLE GRAPH:
Largest:
directed: false
vcount: 0
edges: {
}
Smallest:
directed: false
vcount: 0
edges: {
}
Index:
directed: false
vcount: 0
edges: {
}

LOOPLESS MULTIGRAPH:
Largest:
directed: false
vcount: 0
edges: {
}
Smallest:
directed: false
vcount: 0
edges: {
}
Index:
directed: false
vcount: 0
edges: {
}

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 0
edges: {
}
Smallest:
directed: false
vcount: 0
edges: {
}
Index:
directed: false
vcount: 0
edges: {
}


( 1 2 2 3 )

SIMPLE GRAPH:
Largest:
directed: false
vcount: 4
edges: {
3 2
3 1
3 0
2 1
}
Smallest:
directed: false
vcount: 4
edges: {
3 0
3 2
2 1
3 1
}
Index:
directed: false
vcount: 4
edges: {
3 0
3 1
2 1
3 2
}

LOOPLESS MULTIGRAPH:
Largest:
directed: false
vcount: 4
edges: {
3 1
3 2
1 0
3 2
}
Smallest:
directed: false
vcount: 4
edges: {
3 0
3 1
2 1
3 2
}
Index:
directed: false
vcount: 4
edges: {
3 0
3 1
2 1
3 2
}

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 4
edges: {
3 1
3 2
1 0
3 2
}
Smallest:
directed: false
vcount: 4
edges: {
3 0
3 1
2 1
3 2
}
Index:
directed: false
vcount: 4
edges: {
3 0
3 1
2 1
3 2
}


( 1 2 2 -3 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPY MULTIGRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical


( 1 1 2 3 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPY MULTIGRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical


( 1 2 3 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest:
directed: false
vcount: 3
edges: {
2 1
2 0
2 1
}
Smallest:
directed: false
vcount: 3
edges: {
2 0
2 1
2 1
}
Index:
directed: false
vcount: 3
edges: {
2 0
2 1
2 1
}

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 3
edges: {
2 1
2 0
2 1
}
Smallest:
directed: false
vcount: 3
edges: {
2 0
2 1
2 1
}
Index:
directed: false
vcount: 3
edges: {
2 0
2 1
2 1
}


( 4 4 4 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest:
directed: false
vcount: 3
edges: {
1 0
2 1
2 0
2 1
2 0
1 0
}
Smallest:
directed: false
vcount: 3
edges: {
2 0
1 0
2 0
1 0
2 1
2 1
}
Index:
directed: false
vcount: 3
edges: {
1 0
2 0
2 0
1 0
2 1
2 1
}

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 3
edges: {
1 0
2 1
2 0
2 1
2 0
1 0
}
Smallest:
directed: false
vcount: 3
edges: {
2 0
1 0
2 0
1 0
2 1
2 1
}
Index:
directed: false
vcount: 3
edges: {
1 0
2 0
2 0
1 0
2 1
2 1
}


( 3 5 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 2
edges: {
1 0
1 0
1 0
1 1
}
Smallest:
directed: false
vcount: 2
edges: {
1 0
1 0
1 0
1 1
}
Index:
directed: false
vcount: 2
edges: {
1 0
1 0
1 0
1 1
}


( 5 3 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 2
edges: {
1 0
1 0
1 0
0 0
}
Smallest:
directed: false
vcount: 2
edges: {
1 0
1 0
1 0
0 0
}
Index:
directed: false
vcount: 2
edges: {
1 0
1 0
1 0
0 0
}


( 1 3 3 4 1 2 1 1 1 3 )

SIMPLE GRAPH:
Largest:
directed: false
vcount: 10
edges: {
9 3
3 2
3 1
5 3
9 2
9 1
2 1
8 5
7 6
4 0
}
Smallest:
directed: false
vcount: 10
edges: {
8 3
7 3
6 1
4 2
9 0
9 5
5 2
2 1
9 3
3 1
}
Index:
directed: false
vcount: 10
edges: {
3 0
3 1
2 1
9 1
3 2
9 2
5 3
5 4
9 6
8 7
}

LOOPLESS MULTIGRAPH:
Largest:
directed: false
vcount: 10
edges: {
3 1
9 2
5 3
9 1
3 2
4 0
7 6
8 5
9 1
3 2
}
Smallest:
directed: false
vcount: 10
edges: {
8 3
7 1
6 2
9 4
3 0
5 3
5 1
2 1
9 2
9 3
}
Index:
directed: false
vcount: 10
edges: {
3 0
3 1
2 1
9 1
9 2
3 2
5 3
5 4
9 6
8 7
}

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 10
edges: {
3 1
9 2
5 3
9 1
3 2
4 0
7 6
8 5
9 1
3 2
}
Smallest:
directed: false
vcount: 10
edges: {
8 3
7 1
6 2
9 4
3 0
5 3
5 1
2 1
9 2
9 3
}
Index:
directed: false
vcount: 10
edges: {
3 0
3 1
2 1
9 1
9 2
3 2
5 3
5 4
9 6
8 7
}


( 2 0 3 2 2 2 2 3 )

SIMPLE GRAPH:
Largest:
directed: false
vcount: 8
edges: {
7 2
7 6
7 5
4 2
3 2
4 0
3 0
6 5
}
Smallest:
directed: false
vcount: 8
edges: {
6 2
7 6
5 2
7 5
7 0
3 2
4 0
4 3
}
Index:
directed: false
vcount: 8
edges: {
2 0
7 0
7 2
3 2
4 3
5 4
6 5
7 6
}

LOOPLESS MULTIGRAPH:
Largest:
directed: false
vcount: 8
edges: {
7 2
3 0
5 4
7 6
3 2
5 0
7 4
6 2
}
Smallest:
directed: false
vcount: 8
edges: {
6 2
7 6
7 0
3 0
4 3
5 4
5 2
7 2
}
Index:
directed: false
vcount: 8
edges: {
2 0
7 0
7 2
3 2
4 3
5 4
6 5
7 6
}

LOOPY MULTIGRAPH:
Largest:
directed: false
vcount: 8
edges: {
7 2
3 0
5 4
7 6
3 2
5 0
7 4
6 2
}
Smallest:
directed: false
vcount: 8
edges: {
6 2
7 6
7 0
3 0
4 3
5 4
5 2
7 2
}
Index:
directed: false
vcount: 8
edges: {
2 0
7 0
7 2
3 2
4 3
5 4
6 5
7 6
}


( )
( )

SIMPLE GRAPH:
Largest:
directed: true
vcount: 0
edges: {
}
Smallest:
directed: true
vcount: 0
edges: {
}
Index:
directed: true
vcount: 0
edges: {
}

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 3 0 1 1 1 1 0 1 )
( 2 1 0 2 2 1 0 0 )

SIMPLE GRAPH:
Largest:
directed: true
vcount: 8
edges: {
0 3
0 4
0 5
3 0
4 0
5 4
2 3
7 1
}
Smallest:
directed: true
vcount: 8
edges: {
7 0
2 3
5 4
3 0
0 4
0 3
0 5
4 1
}
Index:
directed: true
vcount: 8
edges: {
0 3
0 4
0 5
2 0
3 4
4 3
5 0
7 1
}

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 3 1 2 3 1 2 2 )
( 2 2 1 2 3 2 2 )

SIMPLE GRAPH:
Largest:
directed: true
vcount: 7
edges: {
4 0
3 4
3 5
3 6
1 4
0 1
0 3
0 5
6 2
6 1
2 6
2 3
5 0
5 4
}
Smallest:
directed: true
vcount: 7
edges: {
2 4
2 0
0 3
0 5
0 6
6 4
6 1
1 3
3 5
3 4
3 1
4 6
5 0
5 2
}
Index:
directed: true
vcount: 7
edges: {
0 4
0 3
0 5
1 6
2 4
2 1
3 0
3 6
3 5
4 0
5 4
5 3
6 1
6 2
}

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 0 1 2 )
( 1 0 2 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 2 1 0 )
( 2 0 1 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 1 2 0 )
( 0 2 1 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 0 2 )
( 2 0 )

SIMPLE GRAPH:
Largest: not graphical
Smallest: not graphical
Index: not graphical

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


( 3 3 3 3 )
( 3 3 3 3 )

SIMPLE GRAPH:
Largest:
directed: true
vcount: 4
edges: {
0 1
0 2
0 3
1 0
1 2
1 3
2 0
2 1
2 3
3 0
3 1
3 2
}
Smallest:
directed: true
vcount: 4
edges: {
3 0
3 1
3 2
2 3
2 0
2 1
1 3
1 2
1 0
0 3
0 2
0 1
}
Index:
directed: true
vcount: 4
edges: {
0 1
0 2
0 3
1 0
1 2
1 3
2 0
2 1
2 3
3 0
3 1
3 2
}

LOOPLESS MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented

LOOPY MULTIGRAPH:
Largest: not implemented
Smallest: not implemented
Index: not implemented


